relay-server/src/actors/healthcheck.rs

```diff
@@ -13,47 +13,155 @@ use relay_system::{Controller, Shutdown};
+    ///
+    /// The result of the message does not have to be awaited. The message will be delivered and
+    /// handled regardless. The communication channel with the service is unbounded, so backlogs
+    /// could occur when sending too many messages.
+    ///
+    /// Sending the message can fail with `Err(SendError)` if the service has shut down.
+    pub async fn send(&self, data: T) -> Result<bool, SendError> {
+        let (responder, rx) = oneshot::channel();
+        let message = Message { data, responder };
+        self.tx.send(message).map_err(|_| SendError)?;
+        rx.await.map_err(|_| SendError)
+    }
+}
+
 pub struct Healthcheck {
     is_shutting_down: bool,
     config: Arc<Config>,
 }
 
 impl Healthcheck {
+    /// Returns the [`Addr`] of the [`Healthcheck`] actor.
+    ///
+    /// Prior to using this, the service must be started using [`Healthcheck::start`].
+    ///
+    /// # Panics
+    ///
+    /// Panics if the service was not started using [`Healthcheck::start`] prior to this being used.
+    pub fn from_registry() -> Addr<HealthcheckMessage> {
+        ADDRESS.read().as_ref().unwrap().clone()
+    }
+
+    /// Creates a new instance of the Healthcheck service.
+    ///
+    /// The service does not run. To run the service, use [`start`](Self::start).
     pub fn new(config: Arc<Config>) -> Self {
         Healthcheck {
             is_shutting_down: false,
             config,
         }
     }
-}
 
-impl Actor for Healthcheck {
-    type Context = Context<Self>;
+    async fn handle_is_healthy(&mut self, message: IsHealthy) -> bool {
+        let upstream = UpstreamRelay::from_registry();
 
-    fn started(&mut self, context: &mut Self::Context) {
-        Controller::subscribe(context.address());
-    }
-}
+        if self.config.relay_mode() == RelayMode::Managed {
+            let fut = upstream.send(IsNetworkOutage).compat();
+            tokio::spawn(async move {
+                if let Ok(is_outage) = fut.await {
+                    metric!(gauge(RelayGauges::NetworkOutage) = if is_outage { 1 } else { 0 });
+                }
+            });
+        }
 
-impl Supervised for Healthcheck {}
+        match message {
+            IsHealthy::Liveness => true, // Liveness always returns true
+            IsHealthy::Readiness => {
+                if self.is_shutting_down {
+                    return false;
+                }
 
-impl SystemService for Healthcheck {}
+                if self.config.requires_auth()
+                    && !upstream
+                        .send(IsAuthenticated)
+                        .compat()
+                        .await
+                        .unwrap_or(false)
+                {
+                    return false;
+                }
 
-impl Default for Healthcheck {
-    fn default() -> Self {
-        unimplemented!("register with the SystemRegistry instead")
+                Aggregator::from_registry()
+                    .send(AcceptsMetrics)
```

Comment:
The call to `IsAuthenticated` and `AcceptsMetrics` used to be parallelized. Especially once we have parallelized message handling, that won't be worth the added code complexity, so it's fine to keep this as is.